STYLEGUIDE.adoc

```diff
@@ -0,0 +1,578 @@
+
+=== Use Text
+
+Use Text instead of String unless you have a good reason not to.
+
+==== Reasoning
+
+It’s 2022, use a proper, unicode aware string type instead of a linked list.
+
+=== Derive `Show`, define `Pretty`, use `Pretty` for human-readable output
+
+Always derive `Show`, do not define it manually.
+
+Always use `Pretty` for human-readable output, not `Show`.
+
+Always define `Pretty` explicitly (when you need it). It's okay to delegate to the `Show` instance if you think it's good enough.
+
+==== Reasoning
+
+The derived version of `Show` is always useful as a way of seeing the explicit strucuture of a value as a Haskell value.
+Defining `Show` can mean that this is no longer true, and you can't do a better job than the derived version.
+
+We use `Pretty` always for human-readable output, because it's actually friendly to layout and the derived `Show` instance is not usually human friendly.
+Therefore if you need to produce output for humans, define a `Pretty` instance.
+This will typically need to be hand-written, unless it happens that you can defer to the `Show` instance, e.g. for simple enums `Show` can be fine since it just prints the constructor names.
+
+=== Use the package name as the root of the module name hierarchy
+
+If the package is `foo-bar`, then the modules should all be `FooBar.X`.
+
+==== Reasoning
+
+See “Naming conventions” https://www.haskellforall.com/2021/05/module-organization-guidelines-for.html[here]. We do it slightly differently (“FooBar” rather than “Foo.Bar”), but the main principle is the same.
+
+=== Reading direction
+
+Try to keep a single line *mostly* reading left-to-right or right-to-left.
+
+==== Examples
+
+Bad:
+
+[source,Haskell]
+----
+traverse (\x -> <some long function body) things
+----
+
+Good:
+
+[source,Haskell]
+----
+for xs $ \x -> …
+----
+
+==== Reasoning
+
+Haskell can get quite condensed and hard to read, especially when the reading direction changes frequently. Often there are symmetrical versions of operators like `<=<` and `>=>` or `=<<` and `>>=` that you can switch between to make code easier to read.
+
+=== Avoid orphan instances except where they’re safe
+
+Avoid orphan instances, but don’t worry about it if https://www.michaelpj.com/blog/2020/10/29/your-orphans-are-fine.html[you can be sure that they’re safe].
+
+==== Reasoning
+
+See the blog post.
+
+=== Make data strict by default, use `nothunks` for long-lived data structures
+
+Use `StrictData` for new code; make fields strict unless you have a good reason not to.
+
+For data structures that might live for a long time, use `nothunks` to assert that they don't contain unexpected thunks.
+
+==== Reasoning
+
+A painful lesson of Haskell in production is that space leaks really suck, are a huge pain to track down, and can originate in surprising locations.
+This suggests that it's worth a bit of proactive paranoia: just make things strict as much as possible, in the hopes of squashing any nascent space leaks.
```

Comment:
nascent?

Reply:
shorter than "before it appears" and can be understood from the context, no?

Reply:
Ah, I thought that word was a typo. I didn't know the meaning.